지도 롱 터치 시 마커 추가 기능 추가

A-by-alimelon opened the pull request:
## 구현내용
### [feat]
- 지도를 길게 누르면 마커를 추가할 수 있는 Alert 창 표시
- interactiveMarker 이름 clusteringMarker로 변경
- point -> latlng 좌표변환 함수 추가
- NMGLatLngBounds -> BoundingBox로 생성해주는 extenstion 구현

### 논의사항
- 마커 추가할 때 id 값을 고유하게 넣어 줄 수 있는 id 생성 함수를 넣어줘야 
레이어가 중복으로 찍히고 사라지는 문제가 해결 되는데 다음 pr에 적용하겠습니다 ㅎㅎ
- infoWindow fetch해와서 연결하는 작업도 다음 pr에서 할게요!

### 화면
![마커추가](https://user-images.githubusercontent.com/62557093/101641334-fa9ad300-3a74-11eb-9834-1f257a353fbe.gif)

A-by-alimelon committed:
[feat] handleLongPress 함수 수정 및 clusteringMarker 이름 변경

- handleLongPress에서 pick된 마커 확인
- interactiveMarker 이름 clusteringMarker로 변경

A-by-alimelon committed:
[feat] interactiveMapView projection 함수 추가

- point -> latlng 좌표변환 함수 추가

A-by-alimelon committed:
[feat] 마커 추가 기능 구현

- mapController에 해당 마커의 tileID를 찾아 interator에 넘기는 함수 구현
- interactor에서 presenter와 clusterService로 coordinate와 id 전달
- presenter에서 viewController로 마커 생성 후 전달
- NMGLatLngBounds -> BoundingBox로 생성해주는 extenstion 구현

Seungeon-Kim commented on Project17-C-Map/InteractiveClusteringMap/Map/MapViewController.swift:
```diff
@@ -129,25 +129,24 @@ final class MapViewController: UIViewController {
+        if pressedMarker is LeafNodeMarker {
+            showEditMode()
+        } else if pressedMarker is ClusteringMarker {
+           print("클러스터")
+        } else {
+            addLeafNodeMarker(at: gesture.location(in: interactiveMapView))
+            print("빈 공간")
```
print문은 빼주세여~

A-by-alimelon replied:
네,, 자꾸빼먹네요 ,,ㅎㅎ

Seungeon-Kim commented on Project17-C-Map/InteractiveClusteringMap/Map/MapController.swift:
```diff
@@ -66,13 +70,27 @@ extension MapController: NMFTileCoverHelperDelegate {
+    static let TR: Int = 1
+}
+
+extension NMGLatLngBounds {
+    
+    func makeBoundingBox() -> BoundingBox {
+        let BL = self.boundsLatLngs[Index.BL]
+        let TR = self.boundsLatLngs[Index.TR]
+        
+        let bottomLeft = Coordinate(x: BL.lng, y: BL.lat)
+        let topRight = Coordinate(x: TR.lng, y: TR.lat)
+        
+        return BoundingBox(topRight: topRight, bottomLeft: bottomLeft)
+    }
+    
+}
```
중복함수 extension에 해주신거 좋네요!

Seungeon-Kim replied:
근데 NMGLatLngBounds에서 extension해주면 뭔가 Cooridnate와 BoundingBox가 종속되는 느낌이 들어서 BoundingBox 안에 파라미터로 lat, lng 를 받아서 생성할수 있도록 static 함수를 만들어주면 좋을거 같기도 해요!

A-by-alimelon replied:
boundingBox 객체를 생성해서 return 하는 거라 import NMapsMap 안해주고 싶은데, ,좀 애매한 것 같아요 ㅠ 
추후에 리팩토링 하면서 다시 고려해 보는게 좋을 것 같아요🙇🏻‍♀️

Seungeon-Kim replied:
그러게요 조금 애매하네요. 파라미터를 lat lng로 받으면 Map을 import를 안해도되긴하는데... 그러면 init랑 다를게없네요 ㅋㅋㅋㅋ

SkydevilK commented on Project17-C-Map/InteractiveClusteringMap/CoreData/DataManagable.swift:
```diff
@@ -12,7 +12,7 @@ protocol DataManagable {
     func deleteAll()
     func delete(coordinate: Coordinate)
-    func add(poi: POI)
+    func add(coordinate: Coordinate)
```
insert로 안하시고 add로 왜 하셨나요?

A-by-alimelon replied:
네이밍 통일 해볼게요~!

eunjeongS2 commented on Project17-C-Map/InteractiveClusteringMap/Map/MapInteractor.swift:
```diff
@@ -10,6 +10,7 @@ import Foundation
     
     func fetch(boundingBoxes: [CLong: BoundingBox], zoomLevel: Double)
+    func add(tileId: CLong, coordinate: Coordinate)
     func remove(tileIds: [CLong])
     func delete(coordinate: Coordinate)
```
이부분 통일시키면 좋을 것 같아용

A-by-alimelon replied:
넵 네이밍 통일 시키겠습니다!

rnfxl92 reviewed (approved):
LGTM

SkydevilK commented on Project17-C-Map/InteractiveClusteringMap/Map/MapInteractor.swift:
```diff
@@ -10,6 +10,7 @@ import Foundation
     func fetch(boundingBoxes: [CLong: BoundingBox], zoomLevel: Double)
+    func add(tileId: CLong, coordinate: Coordinate)
     func remove(tileIds: [CLong])
     func delete(coordinate: Coordinate)
```
remove와 delete 가 왜 다르죠?

eunjeongS2 replied:
이번 pr이랑은 상관없는 것 같긴한데
remove와 delete는 ClusterBusinessLogic 말고 다른 프로토콜로 나누는건 어떤가용 ?? 클러스터 로직은 아닌거 같아서용

A-by-alimelon replied:
@eunjeongS2 맞아요! 클러스터라기 보다는 추가 삭제라 추후 리팩토링하면서 같이 바꿔보면 어떨까요😊

beggu84 reviewed (approved):
그 동안 다들 구조를 잘 잡아놔서 그런건지, 깔끔한 코딩을 하셔서 그런건지,
가독성이 좋네요~

이번 기능도 굿.

eunjeongS2 commented on Project17-C-Map/InteractiveClusteringMap/Clustering/QuadTree/QuadTreeClusteringService.swift:
```diff
@@ -19,6 +19,10 @@ final class QuadTreeClusteringService {
         self.treeDataStore = treeDataStore
     }
     
+    func add(coordinate: Coordinate) {
+        treeDataStore.add(coordinate: coordinate)
+    }
+    
     func delete(coordinate: Coordinate) {
         treeDataStore.remove(coordinate: coordinate)
     }
```
저는 interactor에서 treeDataStore에 직접 add랑 remove해주는게 좋을 것 같은데 어떤가용
클러스터링서비스가 할 로직은 아닌 것 같아서요

A-by-alimelon replied:
네 일단은 인터렉터가 데이터 스토어를 들고 있는쪽으로 변경할게요 *^^*

A-by-alimelon replied:
변경했습니다 ㅎㅅㅎ

eunjeongS2 reviewed (commented):
고생하셨습니다 ~~~

A-by-alimelon commented on Project17-C-Map/InteractiveClusteringMap/Map/MapViewController.swift:
```diff
@@ -218,10 +230,10 @@ final class MapViewController: UIViewController {
                                                                                    lng: leafNodeMarker.coordinate.x))
             markerAnimation = AnimationController.leafNodeAnimation(position: markerLayer.position)
             
-        } else if let interactiveMarker = marker as? InteractiveMarker {
-            markerLayer = interactiveMarker.markerLayer
-            markerLayer?.position = interactiveMapView.projectPoint(from: NMGLatLng(lat: interactiveMarker.coordinate.y,
-                                                                                    lng: interactiveMarker.coordinate.x))
+        } else if let clusteringMarker = marker as? ClusteringMarker {
+            markerLayer = clusteringMarker.markerLayer
+            markerLayer?.position = interactiveMapView.projectPoint(from: NMGLatLng(lat: clusteringMarker.coordinate.y,
+                                                                                    lng: clusteringMarker.coordinate.x))
```
여기 marker 안에 animate 함수를 넣는 편이 낫겠네요 !

rnfxl92 replied:
죠습니다!

A-by-alimelon replied:
죠습니다!

A-by-alimelon committed:
[feat] 리뷰 반영

- 함수 이름 변경

A-by-alimelon committed:
[refac] BusinessLogic 분리

- MapInteractor 채택 프로토콜 분리 => Cluster / Data
- MapInterator에서 treeDataStore 직접 관리

kyungpyoda commented:
사진 보니까 OK 버튼이 빨간색이네요? HIG에 안맞지않나용

A-by-alimelon committed:
[refac] 애니메이션 함수 분리

A-by-alimelon commented:
> 사진 보니까 OK 버튼이 빨간색이네요? HIG에 안맞지않나용

헉 그러네요 ,,!! 변경하겠습니다 감삼당(_ _)

A-by-alimelon committed:
[fix] alert 버튼 hig 변경
